### ISSUE TITLE
Golems duplicated (no portals involved)

### ISSUE THREAD

**Issue by MentalMouse** [Azanor/thaumcraft#1501]
I set up a pair of golems (chop and gather, both stone) to cut down a greatwood, and left them there afterwards.  Later, I noticed a stone gather golem derping nearby -- but when I looked back at the work site, the golem there seemed still to be on the chest!  Examining more closely, I found that both golems had been.... quadruplicated.  I'm assuming this was two duplications, because that's what makes sense, but I'm at a loss for the trigger.  Most of the duplicates were in fact standing in the same spot overlaying each other -- only when I pushed the chop golem did I see it had been duplicated too.  They were standing in a patch of savanna, right near the spawn point.  No other golems were duplicated even once, including the older alchemy and sort golems in my base nearby.

**Comment by MalkContent** [comment]
Did you name them? I had a captured Pech Mage duplicate ones for similarly mysterious reasons. Only time that ever happened, too.

**Comment by MentalMouse** [comment]
Nope, no names.  What's more, today it happened again, quadruplication and all.

**Comment by Azanor** [comment]
This must be a vanilla bug of sorts - the golems once spawned are just normal entities without any kind of spawning code or something similar. This is the second strange golem related bug you have reported - is it on the same world instance or server as the bug where their marks got mixed up?

**Comment by Reassembly** [comment]
I'm also having golems spontaneously duplicating. Named and unnamed golems are duplicating for me, but my other nearby tame mobs (wolves/cats/chickens) don't seem to be duplicating.

It seems to be occurring when chunks are reloading - i.e. after returning from a trip to the Nether or Twilight Forest, or venturing far enough away from my base in the Overworld to unload and reload the chunks.

When it happens, it usually happens to several golems in my base at once - i.e. if I spot one duplicated golem, as many as half of the other 20+ golems in my base have also been duplicated. However, when this happens, not *all* of my golems get duplicated. I've yet to find a pattern to the duplicating/non-duplicating golems, but I'll keep watching for signs.

**Comment by Terpo** [comment]
I also have this problem. Sometimes golems seem also to disappear. 
My treefarm had four golems: Chop, Gather, Use and Sorting. Randomly some of the golems were duplicating or disappearing.

**Comment by MalkContent** [comment]
The "disappearing" might be golems getting stuck in newgrown trees and suffocating

**Comment by Terpo** [comment]
I have also other setups where golems should not die. Crop farm for example. There is a fence and three golems (harvest, gather, sort). There is the same problem with disappearing / duplicating sometimes.

**Comment by MalkContent** [comment]
Do you have pumpkins or melons in your crop farm?

**Comment by Terpo** [comment]
No, nothing that can hurt / suffocate golems. Only some Pams Harvestcraft thingys and Natura Berrys (full grown)

**Comment by MentalMouse** [comment]
Azanor:  No, this is a different game from the world where the marks were getting mixed up -- this world was founded in 4.2.2.

**Comment by Azanor** [comment]
Thanks, I'll see if I can get to the bottom of this.

**Comment by Azanor** [comment]
Hmmm, I wonder if this is some sort of chunk boundary issue. Entities are saved on a per-chunk basis and it might be that a golem that just happens to be right on a chunk boundary is being saved in both chunks. When the chunk gets reloaded then both chunks suddenly have a version of the golem.
The question is why is this happening with golems specifically?

**Comment by Reassembly** [comment]
Chunk boundary issues might aggravate the issue, but they're not the sole cause. One of my recently duplicated golems - an alchemy golem that rarely moved from its home spot unless I was alchemizing something that used Aer essentia - was standing at (5, 11) within its chunk.

Also of possible note - that mostly-stationary Aer-essentia golem's duplicates were "superimposed" on the same spot - I didn't notice the duplicates until I noticed the golem had "too many arms". Other of my golems which were more continuously active - the ones in my tree farm, for example - had their duplicates "scattered" more, sometimes even embedded in nearby glass panes/hoppers or on the wrong side of walls/Paving Stones of Warding (places the original golem shouldn't have been able to walk to in the first place).

Also, I hadn't noticed golems outright vanishing as mentioned above, but I appear to have just lost my named Advanced Sorting Golem. (Brian was my *right-hand man*, man. Brian was my *rock*. Sure, I can replace him, but it won't be the same...)

This has become more of a problem recently, and...hrm. Has anybody had this happen *without* FastCraft installed?

**Comment by Terpo** [comment]
Good question, we have also installed FastCraft on our server.

**Comment by psychofad** [comment]
I have seen this issue with horses on my server....seemed to duplicate after a server restart.  We even went as far as to place armor (gold, diamond, and iron) on some of them and when they would duplicate they would also duplicate the armor.  So this is not just affecting Golems.  Fastcraft also installed on that server as well...could this really be the reason?

**Comment by Azanor** [comment]
I wonder - let me do some testing with fastcraft.

**Comment by Azanor** [comment]
Just to check - does this happen only on servers, or in single player worlds as well?

**Comment by Reassembly** [comment]
My duplications are happening in my single-player game, so it's not just a server issue.

On the other hand, I think I remember reading that single-player games still somehow use the same client-server architecture as multiplayer games, but in a purely local sense. But that might be too pedantic...or is it?

I've noticed on occasion, when the "duplication bug" hits, that certain Botania-related entities (Terra Shatterer and spark augments, that I've seen) will usually also have several non-interactable phantom duplicates appear. The differences being that (a) the Botania duplicates are rendered, but are "phantoms", and cannot be picked up, whereas the golems are fully functional, and (b) the Botania duplicates disappear after closing & reloading the world, while the duplicate golems are just as "real" and "permanent" as the original, even after reloading.

This might be useless information, but I think there's a chance that comparing the relevant hunks of code from Thaumcraft and Botania - in light of the same trigger causing two slightly different bugs with high correlation - might help to reveal the true source, if nothing else does.

**Comment by tschallacka** [comment]
We have the same issue on my server, but then with eldritch. We started with one named guardian at spawn and now we have 8 similarly named guardians in our warded glass box.

I also had my sorting golem(the one linked to 400 locations) duplicated which was nice, my stuff got sorted out twice as fast. It seems to happen mostly during frequent nether hops in and out of the nether

We now have a nice collection of "granny's" on our server, watching over our spawn
http://i.imgur.com/nLaeAxL.png

**Comment by Reassembly** [comment]
I may have just had this happen *without* a chunk reload. I cleared out all the duplicate golems triggered by my last trip in/out of the Outer Lands, then proceeded to make a stack of sanitizing soap. When I was about to move on to the next task, I noticed that my sorting golem *and* my flower-to-petal golems were missing, and there were (I think) six extra golems in my tree farm.

The number of golems gained was definitely more than the number of golems lost, so it's unlikely to be a mis-indexing/reassignment issue. I was in my base for the entire time between the last "duplicate sweep" and the current event, so the chunks the golems were located in *probably* would not have been unloaded/reloaded.

...so not just golems, but guardians as well. Is it Thaumcraft's entities that are acting up, or any Forge-mod-added entities, I wonder?

**Comment by Azanor** [comment]
@tschallacka do you also have Fastcraft on your server?

**Comment by tschallacka** [comment]
No, i drag along on the last version of cauldron available. I will switch to fastcraft when you change forge versions.(which i hope you wont)



--- Oorspronkelijk bericht ---

Van: "Azanor" <notifications@github.com>
Verzonden: 16 januari 2015 06:52
Aan: "Azanor/thaumcraft" <thaumcraft@noreply.github.com>
CC: "tschallacka" <mdibbets@outlook.com>
Onderwerp: Re: [thaumcraft] Golems duplicated (no portals involved) (#1501)

@tschallacka do you also have Fastcraft on your server?

---

**Comment by Azanor** [comment]
@Reassembly The only weird thing about my entities is that they use the forge spawn data interface. At least the guardians and golems do. However that interface is just a way for modders to access something that happens with vanilla mobs anyway so I don't see how it could be the cause either. Apart from that my entities rely 100% on vanilla code to spawn and despawn. 
It is interesting however that it seems to happen with "persistent" entities. In other words mobs that don't despawn automatically after a while. Someone mentioned it happening to horses as well and they are also marked as persistent.

**Comment by tschallacka** [comment]
I just remembererd that we had the same with our reverend sheep in the church of the potato at spawn. At some point we had 3 nametagged sheep in the church.

So it might be a vanilla / forge thing.

--- Oorspronkelijk bericht ---

Van: "Azanor" <notifications@github.com>
Verzonden: 16 januari 2015 07:02
Aan: "Azanor/thaumcraft" <thaumcraft@noreply.github.com>
CC: "tschallacka" <mdibbets@outlook.com>
Onderwerp: Re: [thaumcraft] Golems duplicated (no portals involved) (#1501)

@Reassembly The only weird thing about my entities is that they use the forge spawn data interface. At least the guardians and golems do. However that interface is just a way for modders to access something that happens with vanilla mobs anyway so I don't see how it could be the cause either. Apart from that my entities rely 100% on vanilla code to spawn and despawn.
It is interesting however that it seems to happen with "persistent" entities. In other words mobs that don't despawn automatically after a while. Someone mentioned it happening to horses as well and they are also marked as persistent.

---

**Comment by Azanor** [comment]
Throwing this open to the forge github as well in the hope of shedding some light on this.

**Comment by bloodmc** [comment]
@MentalMouse @Reassembly @Terpo Are you using Forge or Cauldron? Also, have you reproduced this issue with only Forge + Thaumcraft installed?

**Comment by Azanor** [comment]
Just FYI: So far I haven't been able to reproduce this at all. I tried in single player and on a server both with and without Fastcraft. I didn't have any other mods installed. I might just have been getting "lucky", but no amount of chunk loading, dimension switching or world-loading could duplicate a golem for me.

**Comment by tschallacka** [comment]
It happens for us like once a week or something. Our eldritch collection is from two months starting with just the one.

On my server only thaumcraft runs and dynmap and coreprotect. Nothing else.

--- Oorspronkelijk bericht ---

Van: "Azanor" <notifications@github.com>
Verzonden: 16 januari 2015 09:28
Aan: "Azanor/thaumcraft" <thaumcraft@noreply.github.com>
CC: "tschallacka" <mdibbets@outlook.com>
Onderwerp: Re: [thaumcraft] Golems duplicated (no portals involved) (#1501)

Just FYI: So far I haven't been able to reproduce this at all. I tried in single player and on a server both with and without Fastcraft. I didn't have any other mods installed. I might just have been getting "lucky", but no amount of chunk loading, dimension switching or world-loading could duplicate a golem for me.

---

**Comment by Terpo** [comment]
We are running a big modpack on a forge (#1235) base. I did not reproduce it in a singleplayer world. 

It also only happens every few days. Yesterday my crop golem disappeared but the gather one duplicated.

**Comment by Reassembly** [comment]
@Azanor @psychofad I'd forgotten that the issue can horses as well. However, I haven't *noticed* wolves/cats multiplying (or my own horses). Perhaps it's just been much *more* noticeable with golems because they're *infrastructure*.

@bloodmc I'm using Forge 1277. I'm about to try to build a minimal test world to see if I can nail down the minimum requirements to trigger this.

**Comment by MentalMouse** [comment]
Azanor etc.  I'm in SSP, using Forge.   Chunk boundary... hmm.  They are within a chunk or two of the world spawn.   I haven't tried setting up a test world yet.  It has continued happening, but only to golems I put there.  (I'm about ready to start distilling surplus chop cores. :) ) I moved the gather golem and his chest 8-10 or so blocks away, and he stopped duplicating.  It's been quadruplicates each time, but I'm not good at checking every load, and as noted previously this isn't visible at a distance unless one of the gather duplicates goes wandering (digressive rant deleted).

**Comment by Darkona** [comment]
Azanor, may I suggest making a little dev-version with logging in the entityconstructing/death/spawn events, etc?
Then maybe give one of these guys with the problem a jar file with a custom logger to try and see if you can find out  what is going on.

**Comment by Reassembly** [comment]
@Darkona @Azanor I was thinking of suggesting something similar. I don't know if there's a way to hook a logging event into a mod-added entity being unloaded-and-saved and loaded-and-activated with its chunk, but if that's possible, comparing the two should at least reveal if the problem is with unloading (and saving too many times) or loading (and reading too many times).

I suspect this just might be a Forge issue rather than a Thaumcraft issue, but given that it seems to happen more often (or more noticeably) with Thaumcraft entities than with vanilla entities, testing using Thaumcraft might be a good idea.

It's occurred to me that one difference between golems and vanilla mobs is that they're more likely than vanilla entities to be *in motion* when a chunk is loaded/unloaded. I think there's a chance that entities which are crossing chunk boundaries while their source/destination chunks are being loaded/unloaded might be at higher risk of being read/written multiple times. (Although that wouldn't explain *stationary* golems being duplicated, or the "scattering" of duplicates *inside nearby partially-solid blocks* (like glass panes and hoppers) that sometimes *but not always* happens.)

I still haven't managed to trigger duplicate golems in my test world, but in my game world, using ChickenChunks to keep the chunks around my base permanently loaded seems to prevent duplication/loss of golems entirely, when jumping dimensions and/or making long trips in the Overworld. So, that's a **workaround** at least, as well as a further sign that this issue is chunk load/unload related.

**Comment by MentalMouse** [comment]
Confirmed it still happens on TC 4.2.3.3.  Same golem, and now only the one:  

I moved the gather golem -- he doesn't seem to get duplicated anymore, nor does the use golem now sharing his chest.  And I upgraded the chop golem to a thaumium golem, which continues to get duplicated.)  The golem in question is stationed within eyeshot of the spawn point, and in arms-reach of the Z-axis (X= -4) -- in fact, he needs to cross the axis to get to some of his trees.
